risc0/zkvm/src/receipt.rs

```diff
@@ -173,45 +170,141 @@ pub struct SessionReceipt {
     pub journal: Vec<u8>,
 }
 
-/// Provide common functionality implemented by all receipt types.
-#[dyn_partial_eq]
-#[typetag::serde(tag = "type")]
-pub trait Receipt: Debug {
+/// An inner receipt can take the form of a collection of [SegmentReceipts] or a
+/// [SuccinctReceipt].
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+pub enum InnerReceipt {
+    /// The [SegmentReceipts].
+    Flat(SegmentReceipts),
+
+    /// The [SuccinctReceipt].
+    Succinct(SuccinctReceipt),
```

Comment:
If I understand correctly, the primary difference between `Flat` and `Succint` is the circuit that produced the result, and `Succinct` will also be used for the outputs of all other recursion predicates in the future. If this is correct, I'd advocate for calling these the `Rv32im` and `Recursion`.
```suggestion
    /// [Rv32imReceipts], resulting from proving with the RISC-V circuit.
    Rv32im(Rv32imReceipts),

    /// [RecursionReceipt], resulting from proving with the recursion circuit.
    Recursion(RecursionReceipt),
```

Reply:
@jbruestle Will we need a different `InnerReceipt` type for union/finalized predicates? My impression was that this would be needed to handle proof sets and resolution.